### PR TITLE
Add generic connection pooling support

### PR DESCRIPTION
**Pull request by jkotalik**
Usage:
```c#
var client = new ClientBuilder(serviceProvider).UseSockets().UseConnectionPooling().Build();
```

Calling DisposeAsync on the pooled connection returns it to the pool. Need to determine whether we should allow people force disposal of a connection rather than pooling.

### CODE REVIEW

File: src/Bedrock.Framework.Experimental/Transports/Pooling/HttpEndPoint.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+                && other.Kind.Equals(Kind) 
+                && other.Host.Equals(Host) 
+                && other.SslHostName.Equals(SslHostName) 
+                && other.ProxyUri.Equals(ProxyUri) 
+                && other.MaxConnections == MaxConnections;
+        }
+
+        public override int GetHashCode()
+        {
+            return base.GetHashCode()
```
Comment (davidfowl) [inline src/Bedrock.Framework.Experimental/Transports/Pooling/HttpEndPoint.cs]: https://docs.microsoft.com/en-us/dotnet/api/system.hashcode.combine?view=dotnet-plat-ext-3.1

File: src/Bedrock.Framework.Experimental/Transports/Pooling/ConnectionPoolingFactory.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+using System.Collections.Concurrent;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+
+namespace Bedrock.Framework
+{
+    public class ConnectionPoolingFactory : IConnectionFactory
```
Comment (davidfowl) [inline src/Bedrock.Framework.Experimental/Transports/Pooling/ConnectionPoolingFactory.cs]: Maybe the pool type goes away and it's just a factory?

File: src/Bedrock.Framework.Experimental/Transports/Pooling/ConnectionPool.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            // Dispose all connections
+            return default;
+        }
+
+        public ValueTask<ConnectionContext> GetConnectionAsync(EndPoint endPoint, CancellationToken cancellationToken = default)
+        {
+            if (!_pool.TryGetValue(endPoint, out var endPointPool))
+            {
+                endPointPool = new EndPointPool(endPoint, _connectionFactory, endPoint is HttpEndPoint ? ((HttpEndPoint)endPoint).MaxConnections : 1);
+                _pool[endPoint] = endPointPool;
```
Comment (gfoidl) [inline src/Bedrock.Framework.Experimental/Transports/Pooling/ConnectionPool.cs]: This is not threadsafe. Use `GetOrAdd` instead.

File: src/Bedrock.Framework.Experimental/Transports/Pooling/EndPointPool.cs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+{
+    internal class EndPointPool : IEndPointPool, IAsyncDisposable
+    {
+        private EndPoint _endPoint;
+        private List<ConnectionContext> _idleConnections;
+        private IConnectionFactory _connectionFactory;
+        private int _maxConnectionCount;
+        private int _connectionCount;
+
+        private object SyncObj = new object();
```
Comment (gfoidl) [inline src/Bedrock.Framework.Experimental/Transports/Pooling/EndPointPool.cs]: Nit: naming? Is there any standard to follow in this project?

File: src/Bedrock.Framework.Experimental/Transports/Pooling/EndPointPool.cs [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+            {
+                return new ValueTask<ConnectionContext>(Task.FromCanceled<ConnectionContext>(cancellationToken));
+            }
+
+            // TODO figure out how to get a connection lifetime via settings.
+            var list = _idleConnections;
+            List<ConnectionContext> _connectionsToDispose = null;
+            async ValueTask<ConnectionContext> DisposeConnectionsAndReturn(List<ConnectionContext> toDispose, ConnectionContext context)
+            {
+                foreach(var disposeContext in toDispose)
```
Comment (gfoidl) [inline src/Bedrock.Framework.Experimental/Transports/Pooling/EndPointPool.cs]: Nit
```suggestion
                foreach (var disposeContext in toDispose)
```

File: src/Bedrock.Framework.Experimental/Client/ClientExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    public static class ClientExtensions
+    {
+        public static ClientBuilder UseConnectionPooling(this ClientBuilder builder)
+        {
+            // TODO right now, this must be called after UseSockets
+            builder.ConnectionFactory = new ConnectionPoolingFactory(new ConnectionPool(builder.ConnectionFactory));
+            return builder;
+        }
+    }
+}
```
Comment (gfoidl) [inline src/Bedrock.Framework.Experimental/Client/ClientExtensions.cs]: Nit: newline, other files as well

File: src/Bedrock.Framework.Experimental/Transports/Pooling/EndPointPool.cs (hunk 4 above)
Comment (MithrilMan) [reply src/Bedrock.Framework.Experimental/Transports/Pooling/EndPointPool.cs]: agree, I opened an issue about code style https://github.com/davidfowl/BedrockFramework/issues/35

File: src/Bedrock.Framework.Experimental/Client/ClientExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
+        
+namespace Bedrock.Framework
+{
+    public static class ClientExtensions
+    {
+        public static ClientBuilder UseConnectionPooling(this ClientBuilder builder)
```
Comment (davidfowl) [inline src/Bedrock.Framework.Experimental/Client/ClientExtensions.cs]: How does one specify the pooling policy? What sort of configurability does the pool have?

File: src/Bedrock.Framework.Experimental/Transports/Pooling/ConnectionPool.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            // Dispose all connections
+            return default;
+        }
+
+        public ValueTask<ConnectionContext> GetConnectionAsync(EndPoint endPoint, CancellationToken cancellationToken = default)
+        {
+            if (!_pool.TryGetValue(endPoint, out var endPointPool))
+            {
+                endPointPool = new EndPointPool(endPoint, _connectionFactory, endPoint is HttpEndPoint ? ((HttpEndPoint)endPoint).MaxConnections : 1);
```
Comment (davidfowl) [inline src/Bedrock.Framework.Experimental/Transports/Pooling/ConnectionPool.cs]: This shouldn't know about HttpEndPoint, those settings, like MaxConnections should either be an interface that the endpoint can implement or should be options passed into the ctor of this pool.

File: src/Bedrock.Framework.Experimental/Transports/Pooling/IConnectionPool.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
+using System.Collections.Concurrent;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+
+namespace Bedrock.Framework
+{
+    public interface IConnectionPool
```
Comment (davidfowl) [inline src/Bedrock.Framework.Experimental/Transports/Pooling/IConnectionPool.cs]: We don't need this interface. After looking at this I think the IConnectionPool should be an IConnectionFactory and there should be a way to plug in a policy for the pool. Similar to the object pool that exists in https://github.com/dotnet/extensions/tree/master/src/ObjectPool/src. For now, we can remove this interface.

File: src/Bedrock.Framework.Experimental/Transports/Pooling/TaskCompletionSourceWithCancellation.cs [hunk 10]
```diff
@@ -0,0 +1,28 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Bedrock.Framework
+{
+    internal class TaskCompletionSourceWithCancellation<T> : TaskCompletionSource<T>
```
Comment (davidfowl) [inline src/Bedrock.Framework.Experimental/Transports/Pooling/TaskCompletionSourceWithCancellation.cs]: Move to infrastructure.

File: src/Bedrock.Framework/Client/Client.cs [hunk 11]
```diff
@@ -18,6 +18,8 @@ public Client(IConnectionFactory connectionFactory, ConnectionDelegate applicati
 
         public async ValueTask<ConnectionContext> ConnectAsync(EndPoint endpoint, CancellationToken cancellationToken = default)
         {
+            // How to do pooling here. Should the factory handle it?
```
Comment (davidfowl) [inline src/Bedrock.Framework/Client/Client.cs]: Is this comment still valid? I see it's plugged int via an extension method on the builder. Then it's transparent.

File: src/Bedrock.Framework/Client/ClientBuilder.cs [hunk 12]
```diff
@@ -18,7 +18,7 @@ public ClientBuilder(IServiceProvider serviceProvider)
 
         internal static object Key { get; } = new object();
 
-        private IConnectionFactory ConnectionFactory { get; set; } = new ThrowConnectionFactory();
+        internal IConnectionFactory ConnectionFactory { get; set; } = new ThrowConnectionFactory();
```
Comment (davidfowl) [inline src/Bedrock.Framework/Client/ClientBuilder.cs]: Undo, there's a method for this already call for this. `UseConnectionFactory(factory => new MyFactory(factory));`

File: src/Bedrock.Framework.Experimental/Client/ClientExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+        {
+            // TODO right now, this must be called after UseSockets
+            builder.ConnectionFactory = new ConnectionPoolingFactory(new ConnectionPool(builder.ConnectionFactory));
```
Comment (davidfowl) [inline src/Bedrock.Framework.Experimental/Client/ClientExtensions.cs]: `builder.UseConnectionFactory(factory => new ConnectionPoolingFactory(factory));` see https://github.com/davidfowl/BedrockFramework/pull/34/files#r365549269

File: src/Bedrock.Framework.Experimental/Transports/Pooling/IConnectionPool.cs (hunk 9 above)
Comment (jkotalik) [reply src/Bedrock.Framework.Experimental/Transports/Pooling/IConnectionPool.cs]: Yeah, I mostly was following your suggestion of having:
```c#
builder.UsePooling(pool);
```

I'm on the same page as you; this should just be a factory. 

For providing a policy, it seems like the runtime has an abstract ObjectPool class, should we do the same for the policy provided for pooling?

File: src/Bedrock.Framework.Experimental/Client/ClientExtensions.cs (hunk 7 above)
Comment (jkotalik) [reply src/Bedrock.Framework.Experimental/Client/ClientExtensions.cs]: I'm not certain how someone will be able to define the pooling policy yet, but as you mentioned before, probably something similar to what ObjectPool does in the runtime.

**Comment by jkotalik** [comment]
Thanks for the feedback everyone, 🆙 📅

File: src/Bedrock.Framework.Experimental/Transports/Pooling/IConnectionPool.cs (hunk 9 above)
Comment (davidfowl) [reply src/Bedrock.Framework.Experimental/Transports/Pooling/IConnectionPool.cs]: Lets design this in an issue outside of this PR, it'll require a few rounds.

File: src/Bedrock.Framework.Experimental/Transports/Pooling/ConnectionPoolingFactory.cs [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+            // Dispose all connections
+            return default;
+        }
+
+        public ValueTask<ConnectionContext> ConnectAsync(EndPoint endPoint, CancellationToken cancellationToken = default)
+        {
+            if (!_pool.TryGetValue(endPoint, out var endPointPool))
+            {
+                endPointPool = new EndPointPool(endPoint, _innerFactory, endPoint is IMaxConnectionFeature maxConnectionEndpoint ? maxConnectionEndpoint.MaxConnections : 1);
+                _pool.GetOrAdd(endPoint, endPointPool);
```
Comment (gfoidl) [inline src/Bedrock.Framework.Experimental/Transports/Pooling/ConnectionPoolingFactory.cs]: The `if` is not necessary here --> https://github.com/davidfowl/BedrockFramework/pull/37 fixes this.

**Pull request merged**